bindings/go/dag/dag.go

```diff
@@ -0,0 +1,328 @@
+// # Modified from https://github.com/kro-run/kro/blob/7e437f2fe159a1e1c59d8eefd2bfa55320df4489/pkg/graph/dag/dag.go under Apache 2.0 License
+//
+// Original License:
+//
+// Copyright 2025 The Kube Resource Orchestrator Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// We would like to thank the authors of kro for their outstanding work on this code.
+
+package dag
+
+import (
+	"cmp"
+	"fmt"
+	"maps"
+	"slices"
+	"sort"
+	"strings"
+)
+
+var ErrSelfReference = fmt.Errorf("self-references are not allowed")
+
+// Vertex represents a node/vertex in a directed acyclic graph.
+type Vertex[T cmp.Ordered] struct {
+	// ID is a unique identifier for the node
+	ID T
+	// Attributes stores the attributes of the node, such as the component
+	// descriptor.
+	Attributes map[string]any
+	// Edges stores the IDs of the nodes that this node has an outgoing edge to.
+	// In kro, this would be the children of a resource.
+	Edges map[T]map[string]any
+}
+
+// DirectedAcyclicGraph represents a directed acyclic graph.
+type DirectedAcyclicGraph[T cmp.Ordered] struct {
+	// Vertices stores the nodes in the graph
+	Vertices  map[T]*Vertex[T]
+	OutDegree map[T]int // In-degree of each vertex (number of outgoing edges)
+	InDegree  map[T]int // Out-degree of each vertex (number of incoming edges)
```

Comment:
Are the comments correct?

Reply:
Erhhhhhh no I think I mixed them up, mb :D thanks for the hint